### PR TITLE
Particle mesh depictor

### PR DESCRIPTION
**Pull request by brisvag**
This PR implements a way of depicting particles as isosurfaces meshes. Performance is quite bad for medium to high numbers of triangles, but it's a working prototype.

To use it, one needs to have a `volume` (array, xarray or ImageBlock) and pass it as such:
```python
peeper.show(mode='mesh', volume=volume, iso=0.121, step_size=7)
```

### CODE REVIEW

File: peepingtom/datablocks/multiblocks/particleblock.py [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 from .orientedpointblock import OrientedPointBlock
 from ..simpleblocks import PropertyBlock
-from ...depictors import ParticleDepictor
+from ...depictors import ParticlePointDepictor, ParticleMeshDepictor
```
Comment (alisterburt) [inline peepingtom/datablocks/multiblocks/particleblock.py:5]: maybe `ParticleBasisDepictor`?

File: peepingtom/datablocks/simpleblocks/orientationblock.py [hunk 2]
```diff
@@ -69,5 +69,10 @@ def zyx_vectors(self):
             all_vectors[ax] = (self.oriented_vectors(ax).sel(spatial=list(axes)))
         return all_vectors
 
+    @property
+    def zyx(self):
+        zyx = list('xyz')
+        return self.data.sel(spatial=zyx, spatial2=zyx)
```
Comment (alisterburt) [inline peepingtom/datablocks/simpleblocks/orientationblock.py:75]: what's going on here? should it be 'zyx' in the zyx variable?

File: peepingtom/depictors/napari/particledepictor.py [hunk 3]
```diff
@@ -90,3 +98,65 @@ def update(self):
...
+        all_vertices, all_faces = self.tile_mesh(vertices, faces)
+
+        self._make_surface_layer(all_vertices,
+                                 all_faces,
+                                 name=f'{self.name} - particles',
+                                 scale=self.datablock.pixel_size,
+                                 # TODO: color
+                                 )
+
+    def make_volume_mesh(self):
```
Comment (alisterburt) [inline peepingtom/depictors/napari/particledepictor.py:126]: `mesh_volume`?

File: peepingtom/depictors/napari/particledepictor.py [hunk 4]
```diff
@@ -90,3 +98,65 @@ def update(self):
...
+        vertices, faces, _, _ = marching_cubes(self.volume.copy(), level=self.iso, step_size=self.step_size)  # need copy for readonly issues
+        return vertices, faces
+
+    def tile_mesh(self, vertices, faces):
+        ori = self.datablock.orientations.data.values
+
+        # find rotated mesh vertices for each particle
+        all_vertices_rotated = (vertices @ ori)
```
Comment (alisterburt) [inline peepingtom/depictors/napari/particledepictor.py:134]: this reads strange mathematically, usually we apply a transform on a (column) vector by premultiplication
I think vertices should be reshaped to be column vectors (n, 3, 1) and this should be ori @ vertices

**Review by alisterburt — CHANGES_REQUESTED** [review]
generally this is great - I raised a few points in the comments

If the performance is so bad that this isn't actually usable then we may want to wait before merging, what do you think?

File: peepingtom/datablocks/multiblocks/particleblock.py (hunk 1 above)
Comment (brisvag) [reply peepingtom/datablocks/multiblocks/particleblock.py:5]: What would be `Basis`?

File: peepingtom/datablocks/simpleblocks/orientationblock.py (hunk 2 above)
Comment (brisvag) [reply peepingtom/datablocks/simpleblocks/orientationblock.py:75]: You need to provide a list with the oorder of coordinates. Making the list and then passing the variable was less clunky and more legible than `['z', 'y', 'x']`.

File: peepingtom/depictors/napari/particledepictor.py (hunk 3 above)
Comment (brisvag) [reply peepingtom/depictors/napari/particledepictor.py:124]: Can you use mesh as a verb in this context? o.o

File: peepingtom/depictors/napari/particledepictor.py (hunk 4 above)
Comment (brisvag) [reply peepingtom/depictors/napari/particledepictor.py:132]: I'm not entirely shure on this whole section of math... If you have suggestions like this, I'd love to see a PR with cleaner math :P

**Comment by brisvag** [comment]
Yes, I agree that we should wait! Let's leave this PR around, as it gives us a baseline of looks and performance.

**Comment by brisvag** [comment]
Oops, that was not on purpose...

**Pull request closed without merging**